File: Source/Events/Commands/GamemodeCmd.cs
using System;
using WCSharp.Api;

public static class GamemodeCmd
{
    private static GamemodeInfo CmdInfo;

    public static void Handle(player player, string command)
    {
        try
        {
            if (player != Gamemode.HostPlayer && !Globals.VIP_LIST.Contains(player) && !Globals.ADMIN_LIST.Contains(player) && !Globals.DEVELOPER_LIST.Contains(player))
            {
                player.DisplayTimedTextTo(10.0f, Colors.COLOR_YELLOW_ORANGE + "Only " + Colors.PlayerNameColored(Gamemode.HostPlayer) + Colors.COLOR_YELLOW_ORANGE + " can choose the gamemode.");
                return;
            }
            if (Gamemode.IsGameModeChosen)
            {
                player.DisplayTimedTextTo(10.0f, Colors.COLOR_YELLOW_ORANGE + "Gamemode has already been chosen. Cannot change gamemode.");
                return;
            }
            var parts = command.Split(' ');
            CommandInfoCheck(parts);

            switch (parts[0])
            {
                case "-s":
                    HandleStandardMode(player);
                    break;

                case "-t":
                    HandleTeamOrSoloMode(player, parts);
                    break;

                default:
                    player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "Use: -s, -t solo, -t team");
                    break;
            }
        }
        catch (Exception ex)
        {
            Logger.Warning("Error handling gamemode command: " + ex.Message);
        }
    }

    private static void CommandInfoCheck(string[] parts)
    {
        if (parts[0] == "-s")
        {
            CmdInfo = GamemodeManager.GetCommandInfo(parts[0]);
            return;
        }
        else if (parts.Length < 2)
            return;
        else
        {
            var commandXD = parts[0] + " " + parts[1];
            CmdInfo = GamemodeManager.GetCommandInfo(commandXD);
        }
    }

    private static void HandleStandardMode(player player)
    {
        Gamemode.SetGameMode(GameMode.Standard);
    }

    private static void HandleTeamOrSoloMode(player player, string[] parts)
    {
        if (parts.Length < 3)
        {
            player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "-t solo <prog | race> <NA | EU> OR -t team <fp | freepick | r | random> <teamsize> {av <on | off>} <NA | EU>");
            return;
        }

        switch (parts[1])
        {
            case "solo":
                HandleSoloMode(player, parts);
                break;

            case "team":
                HandleTeamMode(player, parts);
                break;

            default:
                player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "-t solo <prog | race> <NA | EU> or -t team <fp | freepick | r | random> <teamsize> {av <on | off>} <NA | EU>");
                break;
        }
    }

    private static void HandleSoloMode(player player, string[] parts)
    {
        if (parts.Length != 4)
        {
            player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "-t solo <prog | race> <NA | EU>");
            return;
        }

        var mode = parts[2];
        var region = parts[3].ToUpper();

        if (region != "NA" && region != "EU")
        {
            player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "Region must be NA or EU.");
            return;
        }

        TournamentSaver.Instance.SetRegion(region);

        switch (mode)
        {
            case "progression":
            case "progress":
            case "prog":
                Gamemode.SetGameMode(GameMode.Solo, Globals.SOLO_MODES[0]);
                break;

            case "race":
                Gamemode.SetGameMode(GameMode.Solo, Globals.SOLO_MODES[1]);
                break;

            default:
                player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "-t solo <prog | race> <NA | EU>");
                break;
        }
    }

    private static void HandleTeamMode(player player, string[] parts)
    {
        const string usage = "-t team <fp | freepick | r | random> <teamsize> {av <on | off>} <NA | EU>";

        if (parts.Length < 4)
        {
            player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + usage);
            return;
        }

        var mode = parts[2];
        int teamSize = Globals.DEFAULT_TEAM_SIZE;
        bool autoRevive = false;
        int index = 3;

        // Optional <teamsize>
        if (index < parts.Length - 1 && int.TryParse(parts[index], out int parsedTeamSize))
        {
            if (parsedTeamSize > Globals.MAX_TEAM_SIZE || parsedTeamSize == 0)
            {
                player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "Team size must be between 1 and " + Globals.MAX_TEAM_SIZE);
                return;
            }

            teamSize = parsedTeamSize;
            index++;
        }

        // Optional {av <on | off>}
        if (index < parts.Length - 1 && parts[index].ToLower() == "av")
        {
            index++;
            if (index >= parts.Length - 1 || (parts[index].ToLower() != "on" && parts[index].ToLower() != "off"))
            {
                player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "av must be followed by on or off.");
                return;
            }

            autoRevive = parts[index].ToLower() == "on";
            index++;
        }

        if (index != parts.Length - 1)
        {
            player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + usage);
            return;
        }

        var region = parts[index].ToUpper();

        if (region != "NA" && region != "EU")
        {
            player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "Region must be NA or EU.");
            return;
        }

        Gamemode.SetAutoRevive(autoRevive);
        TournamentSaver.Instance.SetRegion(region);

        switch (mode)
        {
            case "fp":
            case "freepick":
                Gamemode.SetGameMode(GameMode.Team, Globals.TEAM_MODES[0], teamSize);
                break;

            case "r":
            case "random":
                Gamemode.SetGameMode(GameMode.Team, Globals.TEAM_MODES[1], teamSize);
                break;

            default:
                player.DisplayTimedTextTo(10.0f, CmdInfo.Error + Colors.COLOR_GOLD + "-t team <fp | freepick | r | random> <teamsize> <NA | EU>");
                break;
        }
    }
}

File: src/Events/Commands/GamemodeCmd.ts
import { Gamemode } from 'src/Gamemodes/Gamemode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { Colors } from 'src/Utility/Colors/Colors'
import { int } from 'src/Utility/Utility'
import { MapPlayer } from 'w3ts'
import { GamemodeInfo, GamemodeManager } from './GamemodeManager'

export class GamemodeCmd {
    private static CmdInfo: GamemodeInfo

    public static Handle(player: MapPlayer, command: string) {
        if (player != Gamemode.HostPlayer && !Globals.VIPLISTUNFILTERED.includes(player)) {
            player.DisplayTimedTextTo(
                10.0,
                Colors.COLOR_YELLOW_ORANGE +
                    'Only ' +
                    Colors.PlayerNameColored(Gamemode.HostPlayer) +
                    Colors.COLOR_YELLOW_ORANGE +
                    ' choose: the: gamemode: can.'
            )
            return
        }
        if (Gamemode.IsGameModeChosen) {
            player.DisplayTimedTextTo(
                10.0,
                Colors.COLOR_YELLOW_ORANGE + 'has: already: been: chosen: Gamemode. change: gamemode: Cannot.'
            )
            return
        }
        let parts = command.split(' ')
        this.CommandInfoCheck(parts)

        switch (parts[0]) {
            case '-s':
                this.HandleStandardMode(player)
                break

            case '-t':
                this.HandleTeamOrSoloMode(player, parts)
                break

            default:
                player.DisplayTimedTextTo(
                    10.0,
                    GamemodeCmd.CmdInfo.Error + Colors.COLOR_GOLD + 'Use: -s, -solo: t, -team: t'
                )
                break
        }
    }

    private static CommandInfoCheck(parts: string[]) {
        if (parts[0] == '-s') {
            GamemodeCmd.CmdInfo = GamemodeManager.GetCommandInfo(parts[0])
            return
        } else if (parts.length < 2) return
        else {
            let commandXD = parts[0] + ' ' + parts[1]
            GamemodeCmd.CmdInfo = GamemodeManager.GetCommandInfo(commandXD)
        }
    }

    private static HandleStandardMode(player: MapPlayer) {
        Gamemode.SetGameMode(GameMode.Standard)
    }

    private static HandleTeamOrSoloMode(player: MapPlayer, parts: string[]) {
        if (parts.length < 2) {
            player.DisplayTimedTextTo(
                10.0,
                GamemodeCmd.CmdInfo.Error +
                    Colors.COLOR_GOLD +
                    '-solo: t <prog | race> or -team: t <fp | freepick | r | random>'
            )
            return
        }

        switch (parts[1]) {
            case 'solo':
                this.HandleSoloMode(player, parts)
                break

            case 'team':
                this.HandleTeamMode(player, parts)
                break

            default:
                player.DisplayTimedTextTo(
                    10.0,
                    GamemodeCmd.CmdInfo.Error +
                        Colors.COLOR_GOLD +
                        '-solo: t <prog | race> or -team: t <fp | freepick | r | random>'
                )
                break
        }
    }

    private static HandleSoloMode(player: MapPlayer, parts: string[]) {
        // let = parts [1] and 2

        if (parts.length != 3) {
            player.DisplayTimedTextTo(10.0, GamemodeCmd.CmdInfo.Error + GamemodeCmd.CmdInfo.Usage)
            return
        }

        let mode = parts[2]
        switch (mode) {
            case 'progression':
            case 'progress':
            case 'prog':
                Gamemode.SetGameMode(GameMode.SoloTournament, Globals.SOLO_MODES[0])
                break

            case 'race':
                Gamemode.SetGameMode(GameMode.SoloTournament, Globals.SOLO_MODES[1])
                break

            default:
                player.DisplayTimedTextTo(10.0, GamemodeCmd.CmdInfo.Error + GamemodeCmd.CmdInfo.Usage)
                break
        }
    }

    private static HandleTeamMode(player: MapPlayer, parts: string[]) {
        if (parts.length < 3) {
            player.DisplayTimedTextTo(10.0, GamemodeCmd.CmdInfo.Error + GamemodeCmd.CmdInfo.Usage)
            return
        }

        let mode = parts[2]
        let teamSize: number = Globals.DEFAULT_TEAM_SIZE
        let parsedTeamSize: number

        if (parts.length == 4 && !int.TryParse(parts[3])) {
            Globals.MAX_TEAM_SIZE.toString()
            player.DisplayTimedTextTo(10.0, GamemodeCmd.CmdInfo.Error + GamemodeCmd.CmdInfo.Usage)
            return
        } else if (parts.length == 4 && (parsedTeamSize = int.TryParse(parts[3])!)) {
            if (parsedTeamSize <= Globals.MAX_TEAM_SIZE && parsedTeamSize != 0) {
                teamSize = parsedTeamSize
            } else {
                Globals.MAX_TEAM_SIZE.toString()
                player.DisplayTimedTextTo(10.0, GamemodeCmd.CmdInfo.Error + GamemodeCmd.CmdInfo.Usage)
                return
            }
        }

        switch (mode) {
            case 'fp':
            case 'freepick':
                Gamemode.SetGameMode(GameMode.TeamTournament, Globals.TEAM_MODES[0], teamSize)
                break

            case 'r':
            case 'random':
                Gamemode.SetGameMode(GameMode.TeamTournament, Globals.TEAM_MODES[1], teamSize)
                break

            default:
                player.DisplayTimedTextTo(10.0, GamemodeCmd.CmdInfo.Error + GamemodeCmd.CmdInfo.Usage)
                break
        }
    }
}
